In [48]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.model_selection import GridSearchCV

In [2]:
data = pd.read_csv('P:/python/forestfires.csv')
data

month  day  FFMC    DMC     DC   ISI  temp  RH  wind  rain  ...  monthfeb  \
0     mar  fri  86.2   26.2   94.3   5.1   8.2  51   6.7   0.0  ...         0   
1     oct  tue  90.6   35.4  669.1   6.7  18.0  33   0.9   0.0  ...         0   
2     oct  sat  90.6   43.7  686.9   6.7  14.6  33   1.3   0.0  ...         0   
3     mar  fri  91.7   33.3   77.5   9.0   8.3  97   4.0   0.2  ...         0   
4     mar  sun  89.3   51.3  102.2   9.6  11.4  99   1.8   0.0  ...         0   
..    ...  ...   ...    ...    ...   ...   ...  ..   ...   ...  ...       ...   
512   aug  sun  81.6   56.7  665.6   1.9  27.8  32   2.7   0.0  ...         0   
513   aug  sun  81.6   56.7  665.6   1.9  21.9  71   5.8   0.0  ...         0   
514   aug  sun  81.6   56.7  665.6   1.9  21.2  70   6.7   0.0  ...         0   
515   aug  sat  94.4  146.0  614.7  11.3  25.6  42   4.0   0.0  ...         0   
516   nov  tue  79.5    3.0  106.7   1.1  11.8  31   4.5   0.0  ...         0   

     monthjan  monthjul  monthjun  monthmar  monthmay  monthnov  monthoct  \
0           0         0         0         1         0         0         0   
1           0         0         0         0         0         0         1   
2           0         0         0         0         0         0         1   
3           0         0         0         1         0         0         0   
4           0         0         0         1         0         0         0   
..        ...       ...       ...       ...       ...       ...       ...   
512         0         0         0         0         0         0         0   
513         0         0         0         0         0         0         0   
514         0         0         0         0         0         0         0   
515         0         0         0         0         0         0         0   
516         0         0         0         0         0         1         0   

     monthsep  size_category  
0           0          small  
1           0          small  
2           0          small  
3           0          small  
4           0          small  
..        ...            ...  
512         0          large  
513         0          large  
514         0          large  
515         0          small  
516         0          small  

[517 rows x 31 columns]

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 517 entries, 0 to 516
Data columns (total 31 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   month          517 non-null    object 
 1   day            517 non-null    object 
 2   FFMC           517 non-null    float64
 3   DMC            517 non-null    float64
 4   DC             517 non-null    float64
 5   ISI            517 non-null    float64
 6   temp           517 non-null    float64
 7   RH             517 non-null    int64  
 8   wind           517 non-null    float64
 9   rain           517 non-null    float64
 10  area           517 non-null    float64
 11  dayfri         517 non-null    int64  
 12  daymon         517 non-null    int64  
 13  daysat         517 non-null    int64  
 14  daysun         517 non-null    int64  
 15  daythu         517 non-null    int64  
 16  daytue         517 non-null    int64  
 17  daywed         517 non-null    int64  
 18  monthapr  

In [4]:
data.describe()

FFMC         DMC          DC         ISI        temp          RH  \
count  517.000000  517.000000  517.000000  517.000000  517.000000  517.000000   
mean    90.644681  110.872340  547.940039    9.021663   18.889168   44.288201   
std      5.520111   64.046482  248.066192    4.559477    5.806625   16.317469   
min     18.700000    1.100000    7.900000    0.000000    2.200000   15.000000   
25%     90.200000   68.600000  437.700000    6.500000   15.500000   33.000000   
50%     91.600000  108.300000  664.200000    8.400000   19.300000   42.000000   
75%     92.900000  142.400000  713.900000   10.800000   22.800000   53.000000   
max     96.200000  291.300000  860.600000   56.100000   33.300000  100.000000   

             wind        rain         area      dayfri  ...    monthdec  \
count  517.000000  517.000000   517.000000  517.000000  ...  517.000000   
mean     4.017602    0.021663    12.847292    0.164410  ...    0.017408   
std      1.791653    0.295959    63.655818    0.371006  ...    0.130913   
min      0.400000    0.000000     0.000000    0.000000  ...    0.000000   
25%      2.700000    0.000000     0.000000    0.000000  ...    0.000000   
50%      4.000000    0.000000     0.520000    0.000000  ...    0.000000   
75%      4.900000    0.000000     6.570000    0.000000  ...    0.000000   
max      9.400000    6.400000  1090.840000    1.000000  ...    1.000000   

         monthfeb    monthjan    monthjul    monthjun    monthmar    monthmay  \
count  517.000000  517.000000  517.000000  517.000000  517.000000  517.000000   
mean     0.038685    0.003868    0.061896    0.032882    0.104449    0.003868   
std      0.193029    0.062137    0.241199    0.178500    0.306138    0.062137   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
75%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

         monthnov    monthoct    monthsep  
count  517.000000  517.000000  517.000000  
mean     0.001934    0.029014    0.332689  
std      0.043980    0.168007    0.471632  
min      0.000000    0.000000    0.000000  
25%      0.000000    0.000000    0.000000  
50%      0.000000    0.000000    0.000000  
75%      0.000000    0.000000    1.000000  
max      1.000000    1.000000    1.000000  

[8 rows x 28 columns]

In [55]:
data1 = data.copy()

In [56]:
data1 = data1.drop(['month', 'day'], axis = 1)
data1.columns

Index(['FFMC', 'DMC', 'DC', 'ISI', 'temp', 'RH', 'wind', 'rain', 'area',
       'dayfri', 'daymon', 'daysat', 'daysun', 'daythu', 'daytue', 'daywed',
       'monthapr', 'monthaug', 'monthdec', 'monthfeb', 'monthjan', 'monthjul',
       'monthjun', 'monthmar', 'monthmay', 'monthnov', 'monthoct', 'monthsep',
       'size_category'],
      dtype='object')

In [9]:
data1[data1.duplicated()]

FFMC    DMC     DC   ISI  temp  RH  wind  rain   area  dayfri  ...  \
53   92.1  111.2  654.1   9.6  20.4  42   4.9   0.0   0.00       0  ...   
100  91.4  142.4  601.4  10.6  19.8  39   5.4   0.0   0.00       0  ...   
215  91.7   35.8   80.8   7.8  17.0  27   4.9   0.0  28.66       0  ...   
303  91.1   94.1  232.1   7.1  19.2  38   4.5   0.0   0.00       1  ...   
426  91.6  248.4  753.8   6.3  20.4  56   2.2   0.0   0.00       0  ...   
461  93.7  231.1  715.1   8.4  18.9  64   4.9   0.0   0.00       0  ...   
501  96.1  181.1  671.2  14.3  21.6  65   4.9   0.8   0.00       0  ...   
508  91.0  166.9  752.6   7.1  25.9  41   3.6   0.0   0.00       1  ...   

     monthfeb  monthjan  monthjul  monthjun  monthmar  monthmay  monthnov  \
53          0         0         0         0         0         0         0   
100         0         0         0         0         0         0         0   
215         0         0         0         0         1         0         0   
303         0         0         0         1         0         0         0   
426         0         0         0         0         0         0         0   
461         0         0         0         0         0         0         0   
501         0         0         0         0         0         0         0   
508         0         0         0         0         0         0         0   

     monthoct  monthsep  size_category  
53          0         0          small  
100         0         0          small  
215         0         0          large  
303         0         0          small  
426         0         0          small  
461         0         0          small  
501         0         0          small  
508         0         0          small  

[8 rows x 29 columns]

In [16]:
data1 = data1.drop(labels = [53,100,215,303,426,461,501,508], axis = 0).reset_index(drop = True)

In [17]:
data1[data1.duplicated()]

Empty DataFrame
Columns: [FFMC, DMC, DC, ISI, temp, RH, wind, rain, area, dayfri, daymon, daysat, daysun, daythu, daytue, daywed, monthapr, monthaug, monthdec, monthfeb, monthjan, monthjul, monthjun, monthmar, monthmay, monthnov, monthoct, monthsep, size_category]
Index: []

[0 rows x 29 columns]

In [19]:
data1.iloc[50:55,:3]

FFMC    DMC     DC
50  92.9  137.0  706.4
51  90.2   99.6  631.2
52  92.1  111.2  654.1
53  91.7  114.3  661.3
54  92.9  137.0  706.4

In [20]:
data1.size_category.value_counts()

small    371
large    138
Name: size_category, dtype: int64

In [22]:
size_category = {'small' : 0, 'large' : 1}
data1['size_category'] = data1['size_category'].map(size_category)

In [23]:
data1['size_category'].value_counts()

0    371
1    138
Name: size_category, dtype: int64

In [24]:
data1.shape

(509, 29)

In [25]:
x = data1.iloc[:,:-1]
y = data1.iloc[:,-1]
print(x.head())
y.head()

   FFMC   DMC     DC  ISI  temp  RH  wind  rain  area  dayfri  ...  monthdec  \
0  86.2  26.2   94.3  5.1   8.2  51   6.7   0.0   0.0       1  ...         0   
1  90.6  35.4  669.1  6.7  18.0  33   0.9   0.0   0.0       0  ...         0   
2  90.6  43.7  686.9  6.7  14.6  33   1.3   0.0   0.0       0  ...         0   
3  91.7  33.3   77.5  9.0   8.3  97   4.0   0.2   0.0       1  ...         0   
4  89.3  51.3  102.2  9.6  11.4  99   1.8   0.0   0.0       0  ...         0   

   monthfeb  monthjan  monthjul  monthjun  monthmar  monthmay  monthnov  \
0         0         0         0         0         1         0         0   
1         0         0         0         0         0         0         0   
2         0         0         0         0         0         0         0   
3         0         0         0         0         1         0         0   
4         0         0         0         0         1         0         0   

   monthoct  monthsep  
0         0         0  
1         1         

0    0
1    0
2    0
3    0
4    0
Name: size_category, dtype: int64

In [27]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.25, random_state = 40)
print(x_train.shape, x_test.shape, y_train.shape, y_test.shape)

(381, 28) (128, 28) (381,) (128,)


### Support Vector Machine model building

In [32]:
model = SVC(kernel='rbf')
model.fit(x_train , y_train)

SVC()

In [33]:
preds = model.predict(x_test)
preds

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int64)

In [36]:
acc = accuracy_score(y_test, preds) * 100
print('accuracy is', acc)

accuracy is 76.5625


In [37]:
confusion_matrix(y_test, preds)

array([[95,  0],
       [30,  3]], dtype=int64)

In [39]:
print('accuracy is', ((95+3)/(95+30+3))*100)

accuracy is 76.5625


In [42]:
# poly
model_p = SVC(kernel='poly')
model_p.fit(x_train , y_train)
preds_p = model_p.predict(x_test)
acc_p = accuracy_score(y_test, preds_p) * 100
print('accuracy is', acc_p)
confusion_matrix(y_test, preds_p)

accuracy is 81.25


array([[95,  0],
       [24,  9]], dtype=int64)

### Model building using Gamma and C 

In [50]:
clf = SVC()
param_grid = [{'kernel':['rbf','linear', 'poly'],'gamma':[50,5,10,0.5],'C':[15,14,13,12,11,10,0.1,0.001] }]
gsv = GridSearchCV(clf, param_grid, cv=10)
gsv.fit(x_train, y_train)

GridSearchCV(cv=10, estimator=SVC(),
             param_grid=[{'C': [15, 14, 13, 12, 11, 10, 0.1, 0.001],
                          'gamma': [50, 5, 10, 0.5],
                          'kernel': ['rbf', 'linear', 'poly']}])

In [51]:
gsv.best_params_ , gsv.best_score_ 

({'C': 15, 'gamma': 50, 'kernel': 'linear'}, 0.9843454790823213)

### Model building using 'Gamma = 50', 'c = 15' and 'kernel = linear'

In [52]:
model_gck = SVC(kernel = 'linear', gamma = 50, C = 15)
model_gck.fit(x_train , y_train)

SVC(C=15, gamma=50, kernel='linear')

In [53]:
preds_gck = model_gck.predict(x_test)
acc_gck = accuracy_score(y_test, preds_gck) * 100
print('accuracy is', acc_gck)
confusion_matrix(y_test, preds_gck)

accuracy is 98.4375


array([[93,  2],
       [ 0, 33]], dtype=int64)

In [54]:
print('model accuracy is ', ((93+33)/(93+2+33))*100)

model accuracy is  98.4375


Frist i have load the dataset. And i have check for null values and there si no null values in data. In the dataset already dummies are created for month and day columns so, i have droped month and day column from the dataset. Now i have check for the duplicates and i have observed that duplicates and i have droped the duplicates from the rows then by using reset index i have rearange all the index. In size_category column there is two Categorie with is small and large and i have changed small as 0 and large as 1. And i have divided the data as input and output by naming input as x and output as y. And i have consider size_Categorie column as output and remaining as input. By using train test split i have divided the data into xtrain, ytrain  xtest, ytest. Then i have build a support vector machine model by using SVC method by passing kernel as 'rbf'. Then i have done predictions and the model accuracy is 76.56%. And i have build model by using kernel as poly and i have done predictions and the model accuracy is 81.25%. Then by using gridsearch i have given kernel as 'rbf','linear', 'poly' and some gamma and c values. And i have understood thatby using c=15 , gamma = 50 and kernel is linear i can get good model. And i have build a model by giving that perameters and i have got model accuracy is 98%.